Advanced Capabilities week 2 day 5 update (#49)
microsoft/win-student-devs: docs/30DaysOfPWA/advanced-capabilities/05.md

# 2.5: Caching your app’s data

**Author: Aaron Gustafson [@AaronGustafson](https://twitter.com/AaronGustafson)**

![image of title and author.](_media/day-05.jpg)

Welcome to day 12 of the [30 Days of PWA](https://aka.ms/learn-pwa/30Days-blog) series! Today's blog post will walk you through how to be more effective and organized when it comes to caching.

We’ll start with a quick recap of caching in a PWA, which [Nitya introduced on day 4](../core-concepts/04.md)…

## Caching basics

> Progressive Web Apps have a lot of control over how they manage resource loading. Much of this power stems from Service Workers, which have the ability to intercept, manipulate, and respond directly to network requests. As a compliment to that, the [Cache API](https://aka.ms/learn-PWA/30Days-2.5/developer.mozilla.org/docs/Web/API/Cache) enables a Service Worker to store and retrieve previously-obtained (or created) Response objects, making it possible to avoid having to go to the network to retrieve long-lived resources like stylesheets and images.

Reviewer: I think the right word is `complement` not `compliment`.

> Progressive Web Apps have a lot of control over how they manage resource loading. Much of this power stems from Service Workers, which have the ability to intercept, manipulate, and respond directly to network requests. As a compliment to that, the [Cache API](https://aka.ms/learn-PWA/30Days-2.5/developer.mozilla.org/docs/Web/API/Cache) enables a Service Worker to store and retrieve previously-obtained (or created) Response objects, making it possible to avoid having to go to the network to retrieve long-lived resources like stylesheets and images.

Reviewer 2: ```suggestion
Progressive Web Apps have a lot of control over how they manage resource loading. Much of this power stems from Service Workers, which have the ability to intercept, manipulate, and respond directly to network requests. As a complement to that, the [Cache API](https://aka.ms/learn-PWA/30Days-2.5/developer.mozilla.org/docs/Web/API/Cache) enables a Service Worker to store and retrieve previously-obtained (or created) Response objects, making it possible to avoid having to go to the network to retrieve long-lived resources like stylesheets and images.
```

Before we tuck in, I want to note that the Cache API is available everywhere, not just within a Service Worker. If you’re only using it within the context of a Service Worker, it’s guaranteed to be available, but if you’re using it elsewhere, you’ll want to test to ensure the feature is available:

```js
if ( "caches" in this ) {
  // Yay! The Cache API is accessible as caches.
}
```

> Here’s a quick run-down on how the Cache API works. Keep in mind that it’s Promise-based. Let’s start with creating/opening a cache:

Reviewer: Suggestion to link the word `Promise` to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises

> Here’s a quick run-down on how the Cache API works. Keep in mind that it’s Promise-based. Let’s start with creating/opening a cache:

Reviewer 2: ```suggestion
Here’s a quick run-down on how the Cache API works. Keep in mind that it’s [Promise](https://developer.mozilla.org/docs/Web/JavaScript/Guide/Using_promises)-based. Let’s start with creating/opening a cache:
```

```js
const app_cache = await caches.open("app");
```

This will create a cache "app" if it doesn’t exist already and then open it. You can then add items to the cache using the `add()`, `addAll()`, or `put()`. Here’s a quick run-down of how these work:

```js
// Add a single Request
app_cache.add("/path/to/resource.ext"); 
  // requests the resource and adds it to the cache
app_cache.add(new Request("/path/to/resource.ext"));
  // does the same thing
app_cache.put("/path/to/resource.ext");

>   // same again, but using put()

Reviewer: I'm a bit confused with the indented comments below the lines of code. I feel like it's usually more natural for comment lines to appear before the lines of code they refer to, not after.

Reviewer 2: I originally had them at the end of the lines, but wrapped them under. I’m fine either way, but I didn’t want to confuse with the overall header there of adding a single Request.

Author: I agree with Patrick. Adjusting these to the end of the line.

Reviewer 3: Also threw me off a bit. I read the header comment as associated with the first add()

> app_cache.add("/path/to/resource.ext"); 
>   // requests the resource and adds it to the cache
> app_cache.add(new Request("/path/to/resource.ext"));
>   // does the same thing
> app_cache.put("/path/to/resource.ext");
>   // same again, but using put()

Author: ```suggestion
app_cache.add("/path/to/resource.ext"); // requests the resource and adds it to the cache
app_cache.add(new Request("/path/to/resource.ext")); // does the same thing
app_cache.put("/path/to/resource.ext"); // same again, but using put()
```

```js
cache.delete("/path/to/generated.json");
```

With the preliminaries out of the way, let’s take a look at how to manage your cached data.

## Getting organized

You can store a lot of data in the browser cache—gigabytes worth in some cases. Just because we can doesn’t mean we should though. After all, at a certain point browser will evict an entire site from memory when space gets tight. Better to not be the resource hog when they come looking to clean up space 😉.

Before we can get more intentional about the amount of data we are caching, it helps to begin categorizing that data. For instance, your PWA will have longer-lived content, such as your CSS, JavaScript, site logo, and maybe your offline page. I like to think of those as *assets* and will create a specific cache for those. Similarly, you’re likely going to be caching Requests for other categories of content like pages, images, and so on. I like to define those categories as individual caches so I can add items to and remove items from them in a more intentional way. I define them up at the top of my Service Worker:

```js

> const version = "v1:";

Reviewer: The idea of using a version string in the cache name is very important in Service Workers, and I feel that this article deserves a section about it. Even just a short one. Something that explains why we usually version our caches, and when do we need to bump the version.

Reviewer 2: I think I’d assumed Nitya would cover that in her intro to caching. I can add a mention.

const sw_caches = {
  assets: {
    name: `${version}assets`
  },
  images: {
    name: `${version}images`
  },
  pages: {
    name: `${version}pages`
  }
};
```

> Here you can see I am using a variable to track my cache <var>version</var> (more on that later) and then I set up an object to define the caches I will be working with: assets, images, and pages. Making each an object may seem like overkill, but there’s a reason for it, which I will get to shortly.

Reviewer 2: ```suggestion
Here you can see I am using a variable to track my cache <var>version</var>. Versioning your cache is a best practice because it allows you to completely remove old (stale) caches completely when you ship new versions of long-lived resources (typically in the Service Worker’s "activate" event).

After establishing the version prefix, I set up an object to define the various caches I will be working with: assets, images, and pages. Making each an object may seem like overkill, but there’s a reason for it, which I will get to shortly.
```

> With this setup in place, I can use the `install` event to cache my assets:

Reviewer: ```suggestion
With this setup in place, I can use the service worker `install` event to cache my assets:
```

```js
const offline_page = "/offline/";
const preinstall = [
  "/favicon.png",
  "/c/default.min.css",
  "/c/advanced.min.css",
  "/j/main.min.js",
  offline_page
];

self.addEventListener( "install", function( event ){
  event.waitUntil(
    caches.open( sw_caches.assets.name )
      .then(function( cache ){
        return cache.addAll( preinstall );
      })
  );
});
```

> Here I’ve defined my offline page’s URL (<var>offline_page</var>) separately because I reference that string elsewhere in the Service Worker. Then I’ve included that URL alongside my PWA’s favicon, and its core CSS and JavaScript as <var>preinstall</var> which, in turn, gets fed into `cache.addAll()` as part of the [Service Worker’s `install` event](../core-concepts/04.md). As the Cache API is Promise-based, you can see that the `install` event (<var>event</var>) is asked to wait for the appropriate cache to be open (in this case <var>sw_caches.assets.name</var>) and for the `addAll()` operation to complete.

Reviewer: TIL about the `<var>` element! I had no idea it even existed 😃
I checked and we don't seem to have any custom styles for it on the site. It just has the default UA italic style. I would suggest sticking to wrapping these in backtick characters so they appear like the rest of our code snippets. That, or add a bit of custom styling for vars.

Reviewer 2: Variable names are typically italic. I’m fine with it as they are not a specific API method or object. @beth-panx can overrule me, of course.

Author: The power I hold! 😈 I think this is fine.

We can use this organization in the `fetch` event as well. I typically use different cache/network recipes for different asset types. Similarly, I can store any of these assets in the most appropriate cache. Here’s an abbreviated example:

```js
self.addEventListener( "fetch", event => {
  // Destination gives us a clue as to the type of resource
  const destination = event.request.destination;
  switch ( destination )
  {
    case "image":
      event.respondWith(
        // check the cache first,
        // fall back to the network
          // and store a copy in 
          // sw_caches.images.name
      );
      break;
    case "document":
      event.respondWith(
        // check the network first
          // and store a copy in
          // sw_caches.pages.name,
        // fall back to the cache
      );
    default:
      event.respondWith(
        // network only
      );
  }
});
```

> This is a very simplified example, but you can see how we can what we know about the Request to make decisions about whether we go to the cache or the network first and where we store copies of the Request for later.

Reviewer: ```suggestion
This is a very simplified example, but you can see how to use the Request information to make decisions about whether we go to the cache or the network first and where we store copies of the Request for later.
```

    name: `${version}images`,
    limit: 50
  },
  pages: {
    name: `${version}pages`,
    limit: 10
  }
};
```

Here I’ve set up hard limits for how many items I want in the image and page caches (50 and 10, respectively). With that in place, we can create a much more methodical cache trimming utility:

```js
function trimCache( cache_name, limit ) {
  caches.open( cache_name )
    .then( cache => {
      cache.keys()
        .then( items => {
          if ( items.length > limit ) {
            (async function(){
              let i = 0,
                  end = items.length - limit;
              while ( i < end ) {
                console.log('deleting item', i, items[i]);
                cache.delete( items[i++] );
              }
            })();
          }
        });
    });
}
```

This function takes two arguments: the name of the cache (<var>cache_name</var>) and the max number of items we want the cache to hold (<var>limit</var>). Here’s what it does:

1. Open the named cache (`caches.open()`), then
2. Get the items in the cache (`cache.keys()`), then
3. Check to see if the number of items exceeds the limit,
4. If it does, figure out how many to remove, and (finally)

> 5. Delete the surplus, starting with the first (oldest) item.

Reviewer: For info, there's no need to actually increment these numbered bullet points. Markdown let's us do:

```
1. Open the ...
1. Get the items ...
1. Check to see ...
```

and it will auto-increment, which is cool because you can way more easily update existing lists (no need to re-number everything).
